README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# Bagged OutlierTrees

<!-- badges: start -->

[![R-CMD-check](https://github.com/RafaJPSantos/bagged.outliertrees/workflows/R-CMD-check/badge.svg)](https://github.com/RafaJPSantos/bagged.outliertrees/actions)
[![CRAN
status](https://www.r-pkg.org/badges/version/bagged.outliertrees)](https://CRAN.R-project.org/package=bagged.outliertrees)
<!-- badges: end -->

Bagged OutlierTrees is an explainable unsupervised outlier detection
method based on an ensemble implementation of the existing OutlierTree
procedure (Cortes, 2020). This implementation takes advantage of
bootstrap aggregating (bagging) to improve robustness by reducing the
possible masking effect and subsequent high variance (similarly to
Isolation Forest), hence the name “Bagged OutlierTrees”.

To learn more about the base procedure OutlierTree (Cortes, 2020),
please refer to
[&lt;arXiv:2001.00636&gt;](https://arxiv.org/abs/2001.00636) (the
corresponding GitHub repository can be found
[here](https://github.com/david-cortes/outliertree)). This repository
and its documentation are heavily based on the latter to ensure
consistency and ease-of-use between the packages.


## Installation

<!-- You can install the released version of bagged.outliertrees from [CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("bagged.outliertrees")
```
-->

You can install the development version of `bagged.outliertrees` from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("RafaJPSantos/bagged.outliertrees")
```

## Example

This is a basic example which shows you how to find outliers in the
[hypothyroid
dataset](http://archive.ics.uci.edu/ml/datasets/thyroid+disease):

``` r
library(bagged.outliertrees)

### example dataset with interesting outliers
data(hypothyroid)

### fit a Bagged OutlierTrees model
model <- bagged.outliertrees(hypothyroid,
  ntrees = 100,
  subsampling_rate = 0.75,
  z_outlier = 5,
  nthreads = 1
)

### use the fitted model to find outliers in the training dataset
outliers <- predict(model,
  newdata = hypothyroid,
  min_outlier_score = 0.5,
  nthreads = 1
)
```

``` r
### print the top-5 outliers in human-readable format
print(outliers, outliers_print = 5)
#> Reporting top 5 outliers [out of 28 found]
#> 
#> row [1438] - suspicious column: [FTI] - suspicious value: [394.495412844037]
#>  distribution: 99.93% <= [294.9661] - [mean: 109.855] - [sd: 30.3889] - [norm. obs: 956]
#> 
#> 
#> row [623] - suspicious column: [age] - suspicious value: [455]
#>  distribution: 99.92% <= [92.03] - [mean: 53.3543] - [sd: 18.9409] - [norm. obs: 956]
#> 
#> 
#> row [745] - suspicious column: [T4U] - suspicious value: [2.12]
#>  distribution: 99.89% <= [1.7222] - [mean: 0.9971] - [sd: 0.1542] - [norm. obs: 700]
#>      [age] > [37.5859] (value: 87)
#> 
#> 
#> row [1425] - suspicious column: [FTI] - suspicious value: [161.290322580645]
#>  distribution: 98.70% <= [104.4645] - [mean: 62.5452] - [sd: 17.6197] - [norm. obs: 89]
#>      [TT4] <= [99.0122] (value: 50)
#> 
#> 
#> row [2110] - suspicious column: [FTI] - suspicious value: [2.38095238095238]
#>  distribution: 99.10% >= [49.6384] - [mean: 93.4009] - [sd: 15.6965] - [norm. obs: 188]
#>      [TT4] <= [112.4091] (value: 2)
```

# References

-   [outliertree](https://github.com/david-cortes/outliertree)
-   Cortes, David. “Explainable outlier detection through decision tree
    conditioning.” arXiv preprint
    [arXiv:2001.00636](https://arxiv.org/abs/2001.00636) (2020).
-   [GritBot software](https://www.rulequest.com/gritbot-info.html)


NEWS.md
## version 1.0.0

---

- Initial version.



R/bagged.outliertrees.R
#' @importFrom outliertree "outlier.tree"
#' @importFrom dplyr "slice_sample"
NULL

#' @title Bagged OutlierTrees
#' @description Fit Bagged OutlierTrees ensemble model to normal data with perhaps some outliers.
#' @param df Data Frame with normal data that might contain some outliers. See details for allowed column types.
#' @param ntrees Controls the ensemble size (i.e. the number of OutlierTrees or bootstrapped training sets).
#' A large value is always recommended to build a robust and stable ensemble.
#' Should be decreased if training is taking too much time.
#' @param subsampling_rate Sub-sampling rate used for bootstrapping.
#' A small rate results in smaller bootstrapped training sets, which should not suffer from the masking effect.
#' This parameter should be adjusted given the size of the training data (perhaps a smaller value for large training data and conversely).
#' @param max_depth Maximum depth of the trees to grow. Can also pass zero, in which case it will only look
#' for outliers with no conditions (i.e. takes each column as a 1-d distribution and looks for outliers in
#' there independently of the values in other columns).
#' @param min_gain Minimum gain that a split has to produce in order to consider it (both in terms of looking
#' for outliers in each branch, and in considering whether to continue branching from them). Note that default
#' value for GritBot is 1e-6, with `gain_as_pct` = `FALSE`, but it's recommended to pass higher values (e.g. 1e-1) when using
#' `gain_as_pct` = `FALSE`.
#' @param z_norm Maximum Z-value (from standard normal distribution) that can be considered as a normal
#' observation. Note that simply having values above this will not automatically flag observations as outliers,
#' nor does it assume that columns follow normal distributions. Also used for categorical and ordinal columns
#' for building approximate confidence intervals of proportions.
#' @param z_outlier Minimum Z-value that can be considered as an outlier. There must be a large gap in the
#' Z-value of the next observation in sorted order to consider it as outlier, given by (z_outlier - z_norm).
#' Decreasing this parameter is likely to result in more observations being flagged as outliers.
#' Ignored for categorical and ordinal columns.
#' @param pct_outliers Approximate max percentage of outliers to expect in a given branch.
#' @param min_size_numeric Minimum size that branches need to have when splitting a numeric column. In order to look for
#' outliers in a given branch for a numeric column, it must have a minimum of twice this number
#' of observations.
#' @param min_size_categ Minimum size that branches need to have when splitting a categorical or ordinal column. In order to
#' look for outliers in a given branch for a categorical, ordinal, or boolean column, it must have a minimum of twice
#' this number of observations.
#' @param categ_split How to produce categorical-by-categorical splits. Options are:
#' \itemize{
#'   \item `"binarize"` : Will binarize the target variable according to whether it's equal to each present category
#'   within it (greater/less for ordinal), and split each binarized variable separately.
#'   \item `"bruteforce"` : Will evaluate each possible binary split of the categories (that is, it evaluates 2^n potential
#'   splits every time). Note that trying this when there are many categories in a column will result
#'   in exponential computation time that might never finish.
#'   \item `"separate"` : Will create one branch per category of the splitting variable (this is how GritBot handles them).
#' }
#' @param categ_outliers How to look for outliers in categorical variables. Options are:
#' \itemize{
#'   \item `"tail"` : Will try to flag outliers if there is a large gap between proportions in sorted order, and this
#'   gap is unexpected given the prior probabilities. Such criteria tends to sometimes flag too many
#'   uninteresting outliers, but is able to detect more cases and recognize outliers when there is no
#'   single dominant category.
#'   \item `"majority"` : Will calculate an equivalent to z-value according to the number of observations that do not
#'   belong to the non-majority class, according to formula '(n-n_maj)/(n * p_prior) < 1/z_outlier^2'.
#'   Such criteria  tends to miss many interesting outliers and will only be able to flag outliers in
#'   large sample sizes. This is the approach used by GritBot.
#' }
#' @param numeric_split How to determine the split point in numeric variables. Options are:
#' \itemize{
#'   \item `"mid"` : Will calculate the midpoint between the largest observation that goes to the '<=' branch and the
#'   smallest observation that goes to the '>' branch.
#'   \item `"raw"` : Will set the split point as the value of the largest observation that goes to the '<=' branch.
#' }
#' This doesn't affect how outliers are determined in the training data passed in `df`, but it does
#' affect the way in which they are presented and the way in which new outliers are detected when
#' using `predict`. `"mid"` is recommended for continuous-valued variables, while `"raw"` will
#' provide more readable explanations for counts data at the expense of perhaps slightly worse
#' generalizability to unseen data.
#' @param cols_ignore Vector containing columns which will not be split, but will be evaluated for usage
#' in splitting other columns. Can pass either a logical (boolean) vector with the same number of columns
#' as `df`, or a character vector of column names (must match with those of `df`).
#' Pass `NULL` to use all columns.
#' @param follow_all Whether to continue branching from each split that meets the size and gain criteria.
#' This will produce exponentially many more branches, and if depth is large, might take forever to finish.
#' Will also produce a lot more spurious outiers. Not recommended.
#' @param gain_as_pct Whether the minimum gain above should be taken in absolute terms, or as a percentage of
#' the standard deviation (for numerical columns) or shannon entropy (for categorical columns). Taking it in
#' absolute terms will prefer making more splits on columns that have a large variance, while taking it as a
#' percentage might be more restrictive on them and might create deeper trees in some columns. For GritBot
#' this parameter would always be `FALSE`. Recommended to pass higher values for `min_gain` when passing `FALSE`
#' here. Not that when `gain_as_pct` = `FALSE`, the results will be sensitive to the scales of variables.
#' @param nthreads Number of parallel threads to use when fitting the model.
#' @return An object with the fitted model that can be used to detect more outliers in new data.
#' @references \itemize{
#'   \item GritBot software: \url{https://www.rulequest.com/gritbot-info.html}
#'   \item Cortes, David. "Explainable outlier detection through decision tree conditioning." arXiv preprint arXiv:2001.00636 (2020).
#' }
#' @seealso \link{predict.bagged.outliertrees} \link{print.bagged.outlieroutputs} \link{hypothyroid}
#' @examples
#' library(bagged.outliertrees)
#'
#' ### example dataset with interesting outliers
#' data(hypothyroid)
#'
#' ### fit a Bagged OutlierTrees model
#' model <- bagged.outliertrees(hypothyroid,
#'   ntrees = 10,
#'   subsampling_rate = 0.75,
#'   z_outlier = 5,
#'   nthreads = 1
#' )
#'
#' ### use the fitted model to find outliers in the training dataset
#' outliers <- predict(model,
#'   newdata = hypothyroid,
#'   min_outlier_score = 0.5,
#'   nthreads = 1
#' )
#'
#' ### print the top-10 outliers in human-readable format
#' print(outliers, outliers_print = 10)
#' @export
bagged.outliertrees <- function(df, ntrees = 100L, subsampling_rate = 0.25, max_depth = 4L, min_gain = 1e-2, z_norm = 2.67, z_outlier = 8.0,
                                pct_outliers = 0.01, min_size_numeric = 25L, min_size_categ = 50L,
                                categ_split = "binarize", categ_outliers = "tail", numeric_split = "raw",
                                cols_ignore = NULL, follow_all = FALSE, gain_as_pct = TRUE,
                                nthreads = parallel::detectCores()) {
  cl <- parallel::makeCluster(nthreads)
  doSNOW::registerDoSNOW(cl)
  on.exit(parallel::stopCluster(cl))

  pb <- utils::txtProgressBar(max = ntrees, style = 3)
  progress <- function(n) utils::setTxtProgressBar(pb, n)
  opts <- list(progress = progress)

  df <- stats::na.omit(df)
  cols <- sapply(df, is.logical)
  df[, cols] <- lapply(df[, cols], as.character)

  bagged_model <- foreach::`%dopar%`(
    foreach::foreach(
      i = 1:ntrees,
      .packages = c("outliertree", "dplyr"),
      .inorder = FALSE,
      .options.snow = opts
    ),
    {
      outlier.tree(slice_sample(df, prop = subsampling_rate, replace = TRUE),
        max_depth = max_depth, min_gain = min_gain, z_norm = z_norm, z_outlier = z_outlier,
        pct_outliers = pct_outliers, min_size_numeric = min_size_numeric, min_size_categ = min_size_categ,
        categ_split = categ_split, categ_outliers = categ_outliers,
        cols_ignore = cols_ignore, follow_all = follow_all, gain_as_pct = gain_as_pct,
        outliers_print = FALSE,
        save_outliers = FALSE
      )
    }
  )

  class(bagged_model) <- "bagged.outliertrees"
  return(bagged_model)
}


R/print.bagged.outlieroutputs.R
#' @title Print outliers in human-readable format
#' @description Pretty-prints outliers as output by the `predict` function from a Bagged OutlierTrees
#' model (as generated by function `bagged.outliertrees`).
#' @param x Outliers as returned by predict method on an object from `bagged.outliertrees`.
#' @param outliers_print Maximum number of outliers to print.
#' @param ... No use.
#' @return The same input `x` that was passed (as `invisible`).
#' @seealso \link{bagged.outliertrees} \link{predict.bagged.outliertrees}
#' @examples
#' library(bagged.outliertrees)
#'
#' ### example dataset with interesting outliers
#' data(hypothyroid)
#'
#' ### fit a Bagged OutlierTrees model
#' model <- bagged.outliertrees(hypothyroid,
#'   ntrees = 10,
#'   subsampling_rate = 0.75,
#'   z_outlier = 5,
#'   nthreads = 1
#' )
#'
#' ### use the fitted model to find outliers in the training dataset
#' outliers <- predict(model,
#'   newdata = hypothyroid,
#'   min_outlier_score = 0.5,
#'   nthreads = 1
#' )
#'
#' ### print the top-10 outliers in human-readable format
#' print(outliers, outliers_print = 10)
#' @export
print.bagged.outlieroutputs <- function(x, outliers_print = 15, ...) {
  outliers_print <- min(outliers_print, length(x))
  cat(sprintf(
    "Reporting top %d outliers [out of %d found]\n\n",
    outliers_print,
    length(x)
  ))

  for (i in names(x[1:outliers_print])) {

    ### print suspicious value
    cat(sprintf("row [%s] - suspicious column: [%s] - ", i, x[[i]]$suspicious_value$column))
    cat(sprintf("suspicious value: [%s]\n", x[[i]]$suspicious_value$value))

    ### print distribution
    if (!is.na(as.numeric(x[[i]]$suspicious_value$value))) {
      if (as.numeric(x[[i]]$suspicious_value$value) > as.numeric(x[[i]]$group_statistics$mean)) {
        cat(sprintf(
          "\tdistribution: %.2f%% <= [%s] - [mean: %s] - [sd: %s] - [norm. obs: %s]\n",
          as.numeric(x[[i]]$group_statistics$pct) * 100.,
          x[[i]]$group_statistics$thr,
          x[[i]]$group_statistics$mean,
          x[[i]]$group_statistics$sd,
          x[[i]]$group_statistics$n_obs
        ))
      } else {
        cat(sprintf(
          "\tdistribution: %.2f%% >= [%s] - [mean: %s] - [sd: %s] - [norm. obs: %s]\n",
          as.numeric(x[[i]]$group_statistics$pct) * 100.,
          x[[i]]$group_statistics$thr,
          x[[i]]$group_statistics$mean,
          x[[i]]$group_statistics$sd,
          x[[i]]$group_statistics$n_obs
        ))
      }
    } else {
      cat(sprintf(
        "\tdistribution: %.2f%% in [%s]\n",
        as.numeric(x[[i]]$group_statistics$pct) * 100.,
        x[[i]]$group_statistics$thr
      ))

      cat(sprintf(
        "\t( [norm. obs: %s] - [prior_prob: %.2f%%] - [next smallest: %.2f%%] )\n",
        x[[i]]$group_statistics$n_obs,
        as.numeric(x[[i]]$group_statistics$sd) * 100.,
        as.numeric(x[[i]]$group_statistics$mean) * 100.
      ))
    }

    ### print conditions
    if (length(x[[i]]$conditions) > 0) {
      conditions <- as.data.frame(x[[i]]$conditions)

      for (j in nrow(conditions)) {
        switch(conditions$comparison[j],
          "is NA" = {
            cat(sprintf("\t\t[%s] is NA\n", conditions$column[j]))
          },
          "<=" = {
            cat(sprintf(
              "\t\t[%s] <= [%s] (value: %s)\n",
              conditions$column[j], conditions$value_comp[j], conditions$value_this[j]
            ))
          },
          ">" = {
            cat(sprintf(
              "\t\t[%s] > [%s] (value: %s)\n",
              conditions$column[j], conditions$value_comp[j], conditions$value_this[j]
            ))
          },
          "=" = {
            cat(sprintf("\t\t[%s] = [%s]\n", conditions$column[j], conditions$value_comp[j]))
          },
          "!=" = {
            cat(sprintf(
              "\t\t[%s] != [%s] (value: %s)\n",
              conditions$column, conditions$value_comp[j], conditions$value_this[j]
            ))
          },
          "in" = {
            cat(sprintf(
              "\t\t[%s] in [%s] (value: %s)\n",
              conditions$column[j], conditions$value_comp[j], conditions$value_this[j]
            ))
          }
        )
      }
    }

    cat("\n\n")
  }
}


R/predict.bagged.outliertrees.R
#' @importFrom dplyr "%>%"
#' @importFrom stats "predict"
#' @importFrom rlist "list.filter"
#' @importFrom data.table "data.table"
#' @importFrom data.table "rbindlist"
#' @importFrom data.table "setnames"
NULL

#' @title Predict method for Bagged OutlierTrees
#' @param object A Bagged OutlierTrees object as returned by `bagged.outliertrees`.
#' @param newdata A Data Frame in which to look for outliers according to the fitted model.
#' @param min_outlier_score Minimum outlier score to use when finding outliers.
#' @param nthreads Number of threads to use when predicting.
#' @param ... No use.
#' @return Will return a list of lists with the outliers and their
#' information (each row is an entry in the first list, with the same names as the rows in the input data
#' frame), which can be printed into a human-readable format after-the-fact through functions
#' `print`.
#' @seealso \link{bagged.outliertrees} \link{print.bagged.outlieroutputs}
#' @examples
#' library(bagged.outliertrees)
#'
#' ### example dataset with interesting outliers
#' data(hypothyroid)
#'
#' ### fit a Bagged OutlierTrees model
#' model <- bagged.outliertrees(hypothyroid,
#'   ntrees = 10,
#'   subsampling_rate = 0.75,
#'   z_outlier = 5,
#'   nthreads = 1
#' )
#'
#' ### use the fitted model to find outliers in the training dataset
#' outliers <- predict(model,
#'   newdata = hypothyroid,
#'   min_outlier_score = 0.5,
#'   nthreads = 1
#' )
#'
#' ### print the top-10 outliers in human-readable format
#' print(outliers, outliers_print = 10)
#' @export
predict.bagged.outliertrees <- function(object, newdata, min_outlier_score = 0.95, nthreads = parallel::detectCores(), ...) {
  suppressWarnings({
    i <- outlier_score <- row_name <- outliertree_id <- condition.column <- condition.value_this <- condition.comparison <- suspicious.column <- suspicious.value <- n <- suspicious.column_agg <- statistics.thr <- statistics.pct <- statistics.mean <- statistics.sd <- statistics.n_obs <- condition.value_comp <- NULL

    cl <- parallel::makeCluster(nthreads)
    doSNOW::registerDoSNOW(cl)
    on.exit(parallel::stopCluster(cl))

    pb <- utils::txtProgressBar(max = length(object), style = 3)
    progress <- function(n) utils::setTxtProgressBar(pb, n)
    opts <- list(progress = progress)


    newdata <- stats::na.omit(newdata)
    cols <- sapply(newdata, is.logical)
    newdata[, cols] <- lapply(newdata[, cols], as.character)

    bagged_predictions <- foreach::`%dopar%`(
      foreach::foreach(
        i = 1:length(object),
        .inorder = FALSE,
        .packages = c("outliertree", "rlist", "data.table"),
        .combine = "rbind",
        .options.snow = opts
      ),
      {
        predictions <- list.filter(
          predict(object[[i]], newdata, return_outliers = TRUE, outliers_print = FALSE),
          !is.na(outlier_score)
        )

        predictions_df <- c()
        for (j in names(predictions)) {
          ifelse(length(predictions[[j]]$conditions) > 0,
                 conditions <- rbindlist(predictions[[j]]$conditions, fill = TRUE),
                 conditions <- data.table("column" = NA, "value_this" = NA, "comparison" = NA, "value_comp" = NA)
          )
          setnames(
            conditions, c("column", "value_this", "comparison", "value_comp"),
            c("condition.column", "condition.value_this", "condition.comparison", "condition.value_comp"),
            skip_absent = TRUE
          )

          suspicous_value <- rbindlist(list(predictions[[j]]$suspicous_value))
          setnames(suspicous_value, c("column", "value"), c("suspicious.column", "suspicious.value"), skip_absent = TRUE)

          group_statistics <- rbindlist(list(predictions[[j]]$group_statistics))
          setnames(group_statistics, c("categs_common"), c("statistics.thr"), skip_absent = TRUE)
          setnames(group_statistics, c("categ_maj"), c("statistics.thr"), skip_absent = TRUE)
          setnames(group_statistics, c("pct_common", "pct_next_most_comm", "prior_prob", "n_obs"),
                   c("statistics.pct", "statistics.mean", "statistics.sd", "statistics.n_obs"),
                   skip_absent = TRUE
          )
          setnames(group_statistics, c("mean", "sd", "n_obs"), c("statistics.mean", "statistics.sd", "statistics.n_obs"),
                   skip_absent = TRUE
          )
          setnames(group_statistics, c("upper_thr", "pct_below"), c("statistics.thr", "statistics.pct"), skip_absent = TRUE)
          setnames(group_statistics, c("lower_thr", "pct_above"), c("statistics.thr", "statistics.pct"), skip_absent = TRUE)


          predictions_df <- rbindlist(
            list(
              predictions_df,
              cbind(
                data.table(outliertree_id = i, row_name = j),
                conditions,
                suspicous_value,
                group_statistics
              )
            ),
            fill = TRUE
          )
        }

        return(predictions_df)
      }
    )

    bagged_predictions <- bagged_predictions[, -c("decimals")]

    bagged_predictions <- bagged_predictions %>%

      # Calculate outlier score
      dplyr::group_by(row_name) %>%
      dplyr::mutate(outlier_score = round(dplyr::n_distinct(outliertree_id) / length(object), 2)) %>%

      # Filter minimum outlier score
      dplyr::filter(outlier_score >= min_outlier_score) %>%

      # Only relevant conditions (majority)
      dplyr::add_count(row_name, condition.column, condition.value_this, condition.comparison, suspicious.column, suspicious.value, outlier_score) %>%
      dplyr::filter(n >= (outlier_score * length(object)) / 2 | is.na(condition.column)) %>%

      # Only the most relevant suspicious column
      dplyr::group_by(row_name) %>%
      dplyr::mutate(suspicious.column_agg = paste(suspicious.column, collapse = ", ")) %>%
      dplyr::mutate(suspicious.column_agg = names(sort(table(unlist(strsplit(suspicious.column_agg, ", "))), decreasing = TRUE))[1]) %>%
      dplyr::filter(suspicious.column == suspicious.column_agg) %>%
      dplyr::group_by(row_name) %>%
      dplyr::mutate(statistics.thr = ifelse(!is.na(as.numeric(statistics.thr)),
                                            as.character(round(mean(as.numeric(statistics.thr)), 4)),
                                            paste(sort(unique(statistics.thr)), collapse = ", ")
      )) %>%
      dplyr::mutate(statistics.pct = as.character(round(mean(as.numeric(statistics.pct)), 4))) %>%
      dplyr::mutate(statistics.mean = as.character(round(mean(as.numeric(statistics.mean)), 4))) %>%
      dplyr::mutate(statistics.sd = as.character(round(mean(as.numeric(statistics.sd)), 4))) %>%
      dplyr::mutate(statistics.n_obs = as.character(round(mean(as.numeric(statistics.n_obs)), 0))) %>%

      # Mean statistics
      dplyr::group_by(
        row_name, condition.column, condition.value_this, condition.comparison, suspicious.column, suspicious.value,
        statistics.thr, statistics.pct, statistics.mean, statistics.sd, statistics.n_obs, outlier_score
      ) %>%
      dplyr::summarize(condition.value_comp = paste(condition.value_comp, collapse = ", ")) %>%
      dplyr::rowwise() %>%
      dplyr::mutate(condition.value_comp = ifelse(!is.na(as.numeric(unlist(strsplit(condition.value_comp, ", "))[1])),
                                                  as.character(round(mean(as.numeric(unlist(strsplit(condition.value_comp, ", ")))), 4)),
                                                  paste(sort(unique(unlist(strsplit(condition.value_comp, ", ")))), collapse = ", ")
      ))

    bagged_predictions[bagged_predictions == "NA"] <- NA

    bagged_predictions <- lapply(
      split(bagged_predictions[, -1], bagged_predictions$row_name),
      function(x) {
        list(
          suspicious_value = list(value = unique(x$suspicious.value), column = unique(x$suspicious.column)),
          group_statistics = list(
            thr = unique(x$statistics.thr), pct = unique(x$statistics.pct), mean = unique(x$statistics.mean),
            sd = unique(x$statistics.sd), n_obs = unique(x$statistics.n_obs)
          ),
          conditions = list(
            column = x$condition.column, value_this = x$condition.value_this,
            comparison = x$condition.comparison, value_comp = x$condition.value_comp
          ),
          outlier_score = unique(x$outlier_score)
        )
      }
    )

    bagged_predictions <- bagged_predictions[rlist::list.order(bagged_predictions, -outlier_score)]

    class(bagged_predictions) <- "bagged.outlieroutputs"
    return(bagged_predictions)
  })
}
